apps/emqx_resource/src/emqx_resource_buffer_worker.erl

```diff
@@ -809,6 +816,28 @@ reply_caller_defer_metrics(Id, ?REPLY(ReplyTo, HasBeenSent, Result), QueryOpts)
     end,
     {ShouldAck, PostFn, DeltaCounters}.
 
+%% This is basically used only by rule actions.  To avoid rule action metrics from
+%% becoming inconsistent when we drop messages, we need a way to signal rule engine that
+%% this action has reached a conclusion.
+-spec reply_dropped(reply_fun(), {error, late_reply | request_expired}) -> ok.
+reply_dropped(_ReplyTo = {Fn, Args, #{reply_dropped := true}}, Result) when
+    is_function(Fn), is_list(Args)
+->
+    %% We want to avoid bumping metrics inside the buffer worker, since it's costly.
+    spawn(fun() -> erlang:apply(Fn, Args ++ [Result]) end),
```

Comment:
TBH, this also sounds costly: it's per each dropped message, and their rate tend to rise when resources are already exhausted (i.e. CPU). But I do not have any good suggestions, besides perhaps using `emqx_pool:async_sumbit/2`.

Reply:
Indeed, but I wanted to avoid stressing these workers at least... :sweat_smile: 

Great suggestion, didn't know/remember this function.  :bulb: 

Maybe we could do something similar to the metrics aggregation inside them, and maybe send a message to a specific metrics worker that then evaluates it like `emqx_pool`, or just use `emqx_pool` directly, if it is not already too burdened.  For now, I'll replace this with async_submit, we could later improve this further I guess.